Update __init__.py

luigirovani opened the pull request:
Update of the code for support  to alternative server

fix bug in subtract_from_days

Throwing an exception when the serial cannot be posted to the server, so as not to generate invalid serials

Customization of serial validity time and number of activations

hansalemaos commented:
Thank you! The biggest problem has been so far the inconstant server (and API changes) from transfer. On the weekend, I will check your code and merge it. Is there anything I need to know?

luigirovani commented:
I also had instability problems with the transfer. I looked for several alternative servers, but this also changed compatibility with client-side. And since my software was already serving several users and needed to maintain compatibility, I needed to find a server that met this requeriment.

https://oshi.at/ work well for me, I've been testing it, and I haven't had any issues, except for one day when its certificate was expired. It was still possible to use it over HTTP, but that would require changes on the client side, in addition to the security issues it would cause. On that day, I used Transfer, which was working fine. 

To using Transfer.sh again, just use "oshi=False". I put two functions to upload, uplodad_file_to_oshi and uplodad_file_to_transfer to maintain compatibility with both. Furthermore, I added "raise_for_status()" when request it returns a status code other than 200, in order avoid generate invalid keys.

Both server and client codes will remain compatible with old codes, as the new parameters I added in the code are optional. I tested here using the standard version of client-side and it worked perfectly.

in the future, I will further improve this code. One cool thing about oshi.at is that it also generates an admin link when a file is posted. The link encrypted in the serial is the user link, which allows only one download. The admin link, for example, allows deleting a link, which would be useful in a future version of the code to revoke a license that I want to cancel. This is one of the issues with this license manager because sometimes, I sell a 1-year subscription to the client, but they start using it and give up within 7 days. I refund their money, but the subscription remains valid for 1 year. Soon, I will implement this in the software to be able to revoke the license within 30 days

At last, I found an issue that may occur in the software. Since the library's name (jerryrigserialver) is visible to the software user, and it is open source, a user can simply replace this code with a modified version that always returns that the signature is valid.

Some simple solutions I found to address this:

- Include a couple of vital functions from your program in the library file to create coupling and make the program not work correctly with the tampered version.
- Return a hash of jerryrigserialver to be checked during the check_serial call.

I also achieved source code security using these settings in pyarmor:
pyarmor gen --enable themida --assert-import --assert-call __init__.py

hansalemaos commented:
Nice!! Thank you very much for your help! Yes, the code is 100 % MIT, and always will be. Before I wrote this script, I had been looking for a free / open source license system, but there was nothing (really nothing!) at all. I have read about pyarmor, but haven't gotten into it. The problem is that an experienced programmer can bypass any license protection. Not event the big techs can do anything about it. That's why they are selling software (e.g. Windows) nowadays as a service. I don't know what you are selling, but there are probably no more than 0.1 percent of your users that know how to bypass your license system. So I wouldn't worry much about that. A bigger problem is that the license is still valid after the cancelation of subscription.
